src/ck-core/cklocation.C

```diff
@@ -2423,21 +2420,50 @@ void CkLocCache::requestLocation(CmiUInt8 id)
 
 void CkLocCache::requestLocation(CmiUInt8 id, const int peToTell)
 {
-  if (peToTell == CkMyPe())
+  if (peToTell == CkMyPe()) return;
+
+  LocationMap::const_iterator itr = locMap.find(id);
+  // TODO: If the location is not found, we probably need to buffer this request. Should
+  // only effect very weird corner cases at the moment, and is a problem that already
+  // existed, but should be addressed in the upcoming delivery/buffering cleanup.
+  if (itr != locMap.end())
   {
-    return;
+    thisProxy[peToTell].updateLocation(itr->second);
   }
+}
 
-  int onPe = lastKnown(id);
-  thisProxy[peToTell].updateLocation(id, onPe);
+void CkLocCache::updateLocation(const CkLocEntry& newEntry)
+{
+  CkAssert(newEntry.pe != -1);
+  CkLocEntry& oldEntry = locMap[newEntry.id];
+  if (newEntry.epoch > oldEntry.epoch)
+  {
+    oldEntry = newEntry;
+    notifyListeners(newEntry.id, newEntry.pe);
+  }
 }
 
-void CkLocCache::updateLocation(CmiUInt8 id, int nowOnPe) { inform(id, nowOnPe); }
+void CkLocCache::moveTo(CmiUInt8 id, int pe)
```

Comment:
Minor, but I don't like the nomenclature of this method, I think `moveTo` implies that it's involved in the moving process, it should be called something closer to `updateLocation` in my mind. Perhaps it can even be merged with the existing `updateLocation` (and have the asserts/assumption of existence be checked elsewhere if necessary).

Reply:
Agreed. Both moveTo and insert are a little weird sounding. They are named based on the conditions they are called in rather than what they do to the table which is weird. I'll try to come up with a better name. I'm a bit hesitant to overload updateLocation even more, because frankly the vast majority of operations on the cache are requests or updates, so we'd just end up with a bunch of versions of request and a bunch of versions of update. request/update are also the only entry methods and I was trying to keep them as just entry method calls, not intended for local use, but maybe that's foolish.

Reply:
Given that it's a cache, I think having things mostly being called something to do with requests or updates makes sense. The semantics of different variants should obviously be made clear, but I don't think there's anything inherently bad with multiple functions being called something like `updateLocation` since that's probably the best description of what the function does.

Reply:
Perhaps something like `recordMovement` would be applicable here? "record" doesn't have the side-effecting conations of "update", at least. Maybe increased verbosity could be another option, like, "updateLoc(ation)Map" or "updateLocationEntry"?

As for "insert", "put" is always another option when dealing with maps.

Reply:
Went with recordEmigration. can't say I love it, but can't say I have anything better either.